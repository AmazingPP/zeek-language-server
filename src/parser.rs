use tree_sitter::{Language, Parser, Tree};

extern "C" {
    fn tree_sitter_zeek() -> Language;
}

pub fn parse(source: impl AsRef<[u8]>, old_tree: Option<&Tree>) -> Option<Tree> {
    let language = unsafe { tree_sitter_zeek() };
    let mut parser = Parser::new();
    parser.set_language(language).ok();

    parser.parse(source, old_tree)
}

#[cfg(test)]
mod test {
    use {
        super::parse,
        eyre::{eyre, Result},
        insta::assert_debug_snapshot,
    };

    #[test]
    fn can_parse() -> Result<()> {
        let source = "event zeek_init() {}";
        let tree = parse(&source, None).ok_or_else(|| eyre!("parser returned no tree"))?;
        assert_debug_snapshot!(&tree.root_node().to_sexp());
        Ok(())
    }
}
