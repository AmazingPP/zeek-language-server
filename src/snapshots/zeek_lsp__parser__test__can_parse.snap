---
source: src/parser.rs
assertion_line: 27
expression: "&tree.root_node().to_sexp()"

---
"(source_file (decl (func_decl (func_hdr (event (id) (func_params))) (func_body))))"
